File: core/EditDistance/EditDistance.py
def minDistance(word1: str, word2: str) -> int:
    m, n = len(word1), len(word2)
    if m == 0 or n == 0:
        return max(m, n)

    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(n):
        dp[0][i + 1] = i + 1
    for i in range(m):
        dp[i + 1][0] = i + 1

    for row in range(1, m + 1):
        for col in range(1, n + 1):
            left = dp[row][col - 1] + 1
            down = dp[row - 1][col] + 1
            left_down = dp[row - 1][col - 1]
            if word1[row - 1] != word2[col - 1]:
                left_down += 1

            dp[row][col] = min(min(left, left_down), down)

    return dp[-1][-1]


if __name__ == '__main__':
    print(minDistance(word1 = "intention", word2 = "execution"))

File: core/EditDistance/EditDistance.java
package com.dbc;

import java.util.Arrays;

public class EditDistance {
    public int minDistance(String word1, String word2) {
        if (word1.length() == 0 || word2.length() == 0) return Math.max(word1.length(), word2.length());
        int[][] dp = new int[word1.length() + 1][word2.length() + 1];
        for (int i = 0; i < word1.length() + 1; i++) dp[i][0] = i;
        for (int i = 0; i < word2.length() + 1; i++) dp[0][i] = i;
        for (int i = 1; i < word1.length() + 1; i++) {
            for (int j = 1; j < word2.length() + 1; j++) {
                int left = dp[i][j - 1] + 1;
                int down = dp[i - 1][j] + 1;
                int left_down = dp[i - 1][j - 1];
                if (word1.charAt(i - 1) != word2.charAt(j - 1)) left_down++;
                dp[i][j] = Math.min(left, Math.min(left_down, down));
            }
        }
        return dp[word1.length()][word2.length()];
    }
}
